### ISSUE TITLE
When using copy API for multiple commit runs, cannot persist to volume in image

### ISSUE THREAD

**Issue by erikh** [docker/docker#29303]
**Description**

sorry for the poor title; but basically, if you do this:

Dockerfile (we call this image 'broken' below):

```
FROM debian
VOLUME /var/tmp
```

```
docker create --name cont broken
docker cp foo cont:/var/tmp
docker commit cont broken
docker run -it broken ls /var/tmp
```

No output because the file is not there; it was saved in the volume created by the image directions.

**Steps to reproduce the issue:**
1. See above
2. Test example
3. Watch in wonder

**Additional information you deem important (e.g. issue happens only occasionally):**

I'm happy to write a patch for this, but it seems like it could go a few ways:

* patch to `docker build` to implement the old `NOVOLUME` or `VOLUME []` suggestions to clear volumes from the list, and an equivalent API to do so (I would really prefer this method)
* Changes to `docker cp` or `docker commit` to manage the volume reference in a persistent way
* Changes to `docker create` to explicitly ignore mounting volumes (this is the most elegant, but less practical than the build solution IMO)

**Output of `docker version`:**

1.12.3

**Comment by erikh** [comment]
alternatively, we could just have a way to manipulate runconfig in a way that did it. that would allow both #1 and #3.

**Comment by erikh** [comment]
like, set nil on the volumes map to inherit, and an empty map should not inherit. any other data is an overwrite.

I hope that's sensible.

**Comment by thaJeztah** [comment]
This was proposed/attempted before, see https://github.com/docker/docker/pull/11183, https://github.com/docker/docker/pull/8803, and https://github.com/docker/docker/pull/7133

I'm not sure this could still be changed, and am afraid we're stuck with the current behavior; some users actually use this as a "feature" to exclude files from the final image

/cc @cpuguy83

**Comment by cpuguy83** [comment]
This is all pretty stanky, but I agree having this on create is going to be required (in hostconfig), I think.

**Comment by erikh** [comment]
Yeah. I presumed retaining the original behavior / provide a flag for the new behavior. Sorry if I wasn't clear on that.

From a pure API perspective, I think this is the runconfig (which is `docker/engine-api/types/container.Config` now right? it's been a long time, sorry) which would of course cascade into the hostconfig on create.

As for the API itself, would you prefer:

* a new struct member to toggle this behavior
* an empty array in the Volumes runconfig could trigger this behavior too. a nil could mean inherit. Easier to write, for sure. Not sure if it's more elegant or not, but it plays on the behavior of the volumes config already, at least as I understand how it works.

As for the question of whether the behavior is necessary or not, it's hard to make build tooling for docker without this. I literally have code that looks like this in the latest release:

```go
for _, volume := range Volumes {
  if strings.HasPrefix(mypath, volume) {
    return fmt.Errorf("can't touch this")
  }
}
```

Which I hope we can agree is unacceptable. :)

**Comment by erikh** [comment]
dun dun dun dun

![](https://i.ytimg.com/vi/NDcffJ9wYxI/maxresdefault.jpg)

**Comment by erikh** [comment]
I know it's the holidays, just wanted to ping and hope for the best. :)
